### PR TITLE
contrib/internal/httptrace: record errors in span tags while retrieving client IP

### PR DESCRIPTION
**Pull request by Hellzy**
This change follows the RFC update: https://datadoghq.atlassian.net/wiki/spaces/APS/pages/2118779066/Client+IP+addresses+resolution.
In case multiple IP related headers are found in a request, we need to record information for the backend in the span tags.

Added tags:

- http.request.headers.*
- _dd.multiple-ip-headers

Changes:
- Change the behaviour of `getClientIP` so that it builds the list of IP headers encountered while retrieving the client IP.
- If the list compiled by `getClientIP` holds more than one element, the new tags are set with information from this list
- `genClientIPSpanTags` abstracts this decision logic and returns a map of tag key/values to set for the span
- Add an env var to disable client ip collection altogether

### CODE REVIEW

File: contrib/internal/httptrace/httptrace.go
Comment (knusbaum) [inline contrib/internal/httptrace/httptrace.go:117]: Let's refactor all of this.

The way it's written now, we build one map of header name -> value, then build another of tag name -> value, then finally converts that map to a list of start options for the span. 

Instead, let's just have one function that returns the tag set that can be `append`ed in `StartRequestSpan`.

**Review by knusbaum — CHANGES_REQUESTED** [review]

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go:117]: I think it's good to keep a function which retrieves the IP without worrying about start options, we may want to retrieve the IP for something other than span tags in the future. What about keeping `getClientIP` as is and changing `genClientIPSpanTags` to generate the start options directly. WDYT?

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go:61]: This is to follow the "allow caller to overwrite" convention used in `StartRequestSpan`.

Comment (katiehockman) [reply contrib/internal/httptrace/httptrace.go]: We don't follow that convention on line 64 though, so I'm having a hard time understanding when that convention should be used and when it doesn't need to be.

I also don't like that we're relying on the order of where things fall in a list in order to determine if something can be overridden by the caller. I realize you didn't originally write this code in this PR, but if possible, we should take a look in this PR or a follow-up one to refactor this.

File: contrib/internal/httptrace/httptrace.go
Comment (katiehockman) [inline contrib/internal/httptrace/httptrace.go]: This would be a lot safer with a `strings.Join`. So instead, do something like this:

```
        var headers []string
	for header, ip := range matches {
		tags[ext.HTTPRequestHeaders+"."+header] = ip
		headers = append(headers, fmt.Sprintf("%s:%s", header, ip)
	}
	tags[ext.MultipleIPHeaders] = strings.Join(headers, ",")
	return tags
```

Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: Great, thanks 👍

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go:99]: This isn't valid in the case a single ip header is found, as expected, but that header does not yield a valid IP to return

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: I mean that splitting the part that retrieves the IP from the one that processes it into span tags makes sense to my eyes since retrieving the IP seems like something worth having on its own. I totally agree that doing it all in a single function is possible.

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: I would be in favour of a follow-up PR. Added a task to my queue for this.

Comment (katiehockman) [reply contrib/internal/httptrace/httptrace.go]: Consider filing a Github issue to track that, and adding a TODO here linking to the issue so it's harder to lose track of.

Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: https://github.com/DataDog/dd-trace-go/issues/1352

File: contrib/internal/httptrace/httptrace.go
Comment (knusbaum) [reply contrib/internal/httptrace/httptrace.go]: We should write functions that solve the problems we have. For instance, if we want a list of `ddtrace.StartSpanOption`s that contain tags for the client IP's, it's generally better to have one function that does that than 2 that each do half the job.

Some reasons to break up a function are
1. if that function becomes too large to be comprehensible
2. if you want to reuse parts of the functionality
3. for testing purposes.

There are more, but these are big ones.

Right now, I think reducing the functions will simplify the code, so #1 does not apply, I think.

For no. 2, we may eventually want to use retrieve the IP for something else, but we may not, or we may want it in a different form. Until we actually need that functionality, it doesn't need to be separated. Cleaning this up and simplifying it will make it easier to pull out that logic into its own function later if we ever do need it elsewhere.

For no. 3, you could separate them, but the `getClientIP` functionality will be tested through `genClientIPSpanTags` anyway. If they are merged, that behavior is still easily covered by tests.

Whether or not we have 1 or 2 functions is not as important to me as that the code is simple and easy to understand. Currently the path from headers to tags is convoluted unnecessarily, as they make their way through several stages of maps and transformations, when we really just need one transformation from `*http.Request` to `[]StartSpanOption`.

File: contrib/internal/httptrace/httptrace.go
Comment (knusbaum) [inline contrib/internal/httptrace/httptrace.go]: If there is exactly one IP header, but it is not valid or global, this will return `netaddr.IP{}, nil`. That seems incorrect.

**Review by knusbaum — CHANGES_REQUESTED** [review]
I think there's still a few corner cases here that give strange results and still think this can be greatly simplified.

I'm proposing something more along these lines:
```
// clientIPTags generates the client IP related tags. If there is exactly one, valid IP header, the
// IP will be set in the ext.HTTPClientIP tag. Otherwise, any IP headers will be added as tags with
// ext.HTTPRequestHeaders.<header> keys. If there are no IP headers, the request's RemoteAddr field
// will be used if valid.
func clientIPTags(r *http.Request) []ddtrace.StartSpanOption {
	ipHeaders := defaultIPHeaders
	if len(clientIPHeader) > 0 {
		ipHeaders = []string{clientIPHeader}
	}

	var ips []string
	var tags []ddtrace.StartSpanOption
	for _, hdr := range ipHeaders {
		if v := r.Header.Get(hdr); v != "" {
			ips = append(ips, v)
			tags = append(tags, tracer.Tag(ext.HTTPRequestHeaders+"."+hdr, v))
		}
	}

	if len(ips) == 0 {
		// Only use the server-detected remote address if the client didn't provide any ip headers.
		ips = append(ips, r.RemoteAddr)
	}
	if len(ips) == 1 {
		for _, s := range strings.Split(ips[0], ",") {
			ip := parseIP(strings.TrimSpace(s))
			if ip.IsValid() && isGlobal(ip) {
				return []ddtrace.StartSpanOption{tracer.Tag(ext.HTTPClientIP, ip.String())}
			}
		}
		return tags
	}
	return append(tags, tracer.Tag(ext.MultipleIPHeaders, strings.Join(ips, ",")))
}
```

I still question why we should return the `ext.HTTPRequestHeaders` list only when there is not a valid IP. Is the name of the header not important when the IP is valid?
Similarly, why do we handle `r.RemoteAddr` so differently? It can only ever appear in the `ext.HTTPClientIP` tag, but never anywhere else.

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: Refactored in 5f702117d6, let me know what you think

**Comment by knusbaum** [comment]
@Hellzy My previous review was posted before I saw the new changes. I'll review the new changes once this is ready.

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: The new refactored function has a validity check

File: contrib/internal/httptrace/httptrace.go
Comment (Hellzy) [reply contrib/internal/httptrace/httptrace.go]: Disappeared after refactor. Also, this looks like an old chunk from several versions ago, weird.

**Review by knusbaum — APPROVED** [review]

**Comment by knusbaum** [comment]
Pending approval from @katiehockman

**Comment by Hellzy** [comment]
> Pending approval from @katiehockman

Checked w/ Katie, merging.

**Pull request merged**
